imports.go

```diff
@@ -30,60 +30,66 @@ type ImportedData struct {
 }
 
 // An Importer imports data from a path.
+// TODO(sbarzowski) caching of errors (may require breaking changes)
 type Importer interface {
-	Import(codeDir string, importedPath string) (*ImportedData, error)
+	// Import fetches data from a given path. It may be relative
+	// to the file where we do the import. What "relative path"
+	// means depends on the importer.
+	//
+	// It is required that:
+	// a) for given (from, path) the same (contents, foundAt) are
+	//    returned on subsequent calls.
+	// b) for given foundAt, the contents are always the same
+	//
+	// Note that by "the same contents" we mean the same pointer and
+	// the same underlying data.
+	//
+	// It is recommended that if there are multiple locations that
+	// need to be probed (e.g. relative + multiple library paths)
+	// then all results of all attempts will be cached separately,
+	// both nonexistence and contents of existing ones.
+	// FileImporter may serve as an example.
+	Import(importedFrom, path string) (contents *string, foundAt string, err error)
 }
 
-// ImportCacheValue represents a value in an imported-data cache.
-type ImportCacheValue struct {
-	// nil if we got an error
-	data *ImportedData
-
-	// nil if we got an error or have only imported it via importstr
-	asCode potentialValue
-
-	// Errors can occur during import, we have to cache these too.
-	err error
-}
-
-type importCacheKey struct {
-	dir          string
-	importedPath string
-}
-
-type importCacheMap map[importCacheKey]*ImportCacheValue
-
 // ImportCache represents a cache of imported data.
 type ImportCache struct {
-	cache    importCacheMap
-	importer Importer
+	foundAtVerification map[string]*string
+	codeCache           map[string]potentialValue
+	importer            Importer
 }
 
 // MakeImportCache creates and ImportCache using an importer.
 func MakeImportCache(importer Importer) *ImportCache {
-	return &ImportCache{importer: importer, cache: make(importCacheMap)}
+	return &ImportCache{
+		importer:            importer,
+		foundAtVerification: make(map[string]*string),
+		codeCache:           make(map[string]potentialValue),
+	}
 }
 
-func (cache *ImportCache) importData(key importCacheKey) *ImportCacheValue {
-	if cached, ok := cache.cache[key]; ok {
-		return cached
+func (cache *ImportCache) importData(codePath, importedPath string) (contents *string, foundAt string, err error) {
```

Comment:
Can you be consistent with the parameter names, because in Importer they are `importedFrom, path`